apps/src/netsim/NetSimConnection.js

```diff
@@ -0,0 +1,334 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Handles client connection status with netsim data services
+ */
+
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+var _ = require('../utils').getLodash();
+var netsimStorage = require('./netsimStorage');
+var NetSimLogger = require('./NetSimLogger');
+var LogLevel = NetSimLogger.LogLevel;
+var ObservableEvent = require('./ObservableEvent');
+
+/**
+ * How often a keep-alive message should be sent to the instance lobby
+ * @type {number}
+ * @const
+ */
+var KEEP_ALIVE_INTERVAL_MS = 2000;
+
+/**
+ * Milliseconds before a client is considered 'disconnected' and
+ * can be cleaned up by another client.
+ * @type {number}
+ * @const
+ */
+var CONNECTION_TIMEOUT_MS = 30000;
+
+/**
+ * A connection to a NetSim instance
+ * @param {string} displayName - Name for person on local end
+ * @param {NetSimLogger} logger - A log control interface, default nullimpl
+ * @constructor
+ */
+var NetSimConnection = function (displayName, logger /*=new NetSimLogger(NONE)*/) {
+  /**
+   * Display name for user on local end of connection, to be uploaded to others.
+   * @type {string}
+   * @private
+   */
+  this.displayName_ = displayName;
+
+  /**
+   * Instance of logging API, gives us choke-point control over log output
+   * @type {NetSimLogger}
+   * @private
+   */
+  this.logger_ = logger;
+  if (undefined === this.logger_) {
+    this.logger_ = new NetSimLogger(console, LogLevel.NONE);
+  }
+
+  /**
+   * Access object for instance lobby
+   * @type {netsimStorage.SharedStorageTable}
+   * @private
+   */
+  this.lobbyTable_ = null;
+
+  /**
+   * This connection's unique Row ID within the lobby table.
+   * If undefined, we aren't connected to an instance.
+   * @type {number}
+   */
+  this.myLobbyRowID_ = undefined;
+
+  /**
+   * Allows others to subscribe to connection status changes.
+   * args: none
+   * Notifies on:
+   * - Connect to instance
+   * - Disconnect from instance
+   * @type {ObservableEvent}
+   */
+  this.statusChanges = new ObservableEvent();
+
+  /**
+   * When the next keepAlive update should be sent to the lobby
+   * @type {Number}
+   * @private
+   */
+  this.nextKeepAliveTime_ = Infinity;
+
+  // Bind to onBeforeUnload event to attempt graceful disconnect
+  window.addEventListener('beforeunload', _.bind(this.onBeforeUnload_, this));
```

Comment:
I would suggest using Array.prototype.bind here (e.g. this.onBeforeUnload_.bind(this)), to avoid pulling in lodash. It works on all the browsers we support.